File: surf-api-velocity/surf-api-velocity-server/src/main/java/dev/slne/surf/surfapi/velocity/server/packet/PacketApiLoader.java
package dev.slne.surf.surfapi.velocity.server.packet;

import com.github.retrooper.packetevents.PacketEvents;
import com.velocitypowered.api.plugin.PluginContainer;
import com.velocitypowered.api.proxy.ProxyServer;
import io.github.retrooper.packetevents.velocity.factory.VelocityPacketEventsBuilder;
import org.jetbrains.annotations.ApiStatus;

/**
 * The PacketApiLoader class is responsible for loading and initializing the packet API.
 * It sets up packet events, entity lib, and entity counter.
 */
@ApiStatus.Internal
public final class PacketApiLoader {

    private final ProxyServer server;
    private final PluginContainer pluginContainer;

    /**
     * This class represents a PacketApiLoader.
     *
     * @param server The ProxyServer instance used for loading the packet API.
     * @param pluginContainer The PluginContainer instance used for loading the packet API.
     */
    public PacketApiLoader(ProxyServer server, PluginContainer pluginContainer) {
        this.server = server;
        this.pluginContainer = pluginContainer;
    }

    /**
     * Initializes the necessary components for the onLoad event.
     */
    public void onLoad() {
        setupPacketEvents();
    }

    /**
     * Initializes the plugin by calling the init method of the PacketEvents API.
     */
    public void onEnable() {
        PacketEvents.getAPI().init();
    }

    /**
     * This method is called when the plugin is being disabled.
     * It initializes the PacketEvents API.
     */
    public void onDisable() {
        PacketEvents.getAPI().terminate();
    }

    /**
     * Sets up the packet events by configuring the PacketEvents API.
     * This method should be called during the onLoad event.
     */
    private void setupPacketEvents() {
        PacketEvents.setAPI(VelocityPacketEventsBuilder.build(server, pluginContainer));
        PacketEvents.getAPI().load();
    }
}

File: surf-api-paper/surf-api-paper-server/src/main/kotlin/dev/slne/surf/api/paper/server/packet/PacketApiLoader.kt
package dev.slne.surf.api.paper.server.packet

import com.github.retrooper.packetevents.PacketEvents
import dev.slne.surf.api.core.extensions.packetEvents
import dev.slne.surf.api.paper.event.register
import dev.slne.surf.api.paper.event.unregister
import dev.slne.surf.api.paper.nms.NmsUseWithCaution
import dev.slne.surf.api.paper.nms.common.AbstractChannelInjector
import dev.slne.surf.api.paper.nms.common.NmsProvider
import dev.slne.surf.api.paper.packet.listener.SurfPaperPacketListenerApi
import dev.slne.surf.api.paper.packet.listener.listener.PacketListener
import dev.slne.surf.api.paper.server.packet.lore.PluginDisablePacketLoreListener
import dev.slne.surf.api.paper.server.plugin
import io.github.retrooper.packetevents.factory.spigot.SpigotPacketEventsBuilder

@OptIn(NmsUseWithCaution::class)
object PacketApiLoader {

    private var versionPacketListeners: List<PacketListener> = emptyList()

    fun onLoad() {
        setupPacketEvents()
    }

    @OptIn(NmsUseWithCaution::class)
    fun onEnable() {
        packetEvents.init()

        // Register version-specific packet listeners from NmsProvider
        val provider = NmsProvider.current
        provider.initialize()
        
        versionPacketListeners = provider.createPacketListeners()
        for (listener in versionPacketListeners) {
            SurfPaperPacketListenerApi.registerListeners(listener)
        }

        AbstractChannelInjector.instance.register()
        PluginDisablePacketLoreListener.register()
    }

    @OptIn(NmsUseWithCaution::class)
    fun onDisable() {
        val provider = NmsProvider.current

        packetEvents.terminate()
        for (listener in versionPacketListeners) {
            SurfPaperPacketListenerApi.unregisterListeners(listener)
        }
        versionPacketListeners = emptyList()

        PluginDisablePacketLoreListener.unregister()
        AbstractChannelInjector.instance.unregister()
        provider.shutdown()
    }

    private fun setupPacketEvents() {
        PacketEvents.setAPI(SpigotPacketEventsBuilder.build(plugin))
        packetEvents.load()
    }
}
